### PR TITLE
fix type instability in exp exp1m for Complex{Int}

### LINKED ISSUE #18785: exp(2*im) is type unstable
Returns a `::Union{Complex{Float64},Complex{Int64}}`

### PR DESCRIPTION
**Pull request by KristofferC**
Fixes #18785

### CODE REVIEW

**Review by JeffBezanson — APPROVED** [review]

File: base/complex.jl [hunk 1]
```diff
@@ -424,7 +424,8 @@ function log2(z::Complex)
     a/log(oftype(real(a),2))
 end
 
-function exp(z::Complex)
+exp(z::Complex) = exp(float(z))
```
Comment (yuyichao) [inline base/complex.jl]: Why can't type inference figure this out given `isnan(::Int)` and `isfinite(::Int)` are both constants? I was thinking about doing this a while ago (before we had constant propagation in type inference) but this adds branches (`isfinite` one in particular) that seems hard for LLVM to eliminat

Comment (yuyichao) [inline base/complex.jl]: Ref https://github.com/JuliaLang/julia/pull/13539 for a better workaround to the same issue.

**Review by yuyichao — CHANGES_REQUESTED** [review]

File: base/complex.jl (hunk 1 above)
Comment (yuyichao) [inline base/complex.jl]: And also the (probably) new preferred way of adding similar test one line above the one you added.

Comment (KristofferC) [inline base/complex.jl]: I think I missunderstand you but the test above the one I added looks pretty similar to how I did it?

Comment (yuyichao) [inline base/complex.jl]: I think you need to rebase https://github.com/JuliaLang/julia/blob/fe29cb590d85df1d2c10ccc3c929f38418121f37/test/complex.jl#L945

Comment (KristofferC) [inline base/complex.jl]: Oh yeah, of course. Thanks.

**Comment by KristofferC** [comment]
Subsumed by #18793

**Pull request closed without merging**